ls コマンドを作るオブジェクト指向版のプログラムを作成

sjabcdefin opened the pull request:
ls コマンドを作るオブジェクト指向版のプログラムを作成いたしました。
ご確認お願いいたします。

sjabcdefin committed:
Add: Create ls command program.

オブジェクト指向版のlsコマンド表示プログラムを作成。

sjabcdefin committed:
Update: Refactor print_file_names method, Move formatted_detail method from FileLister to FileDetail and Define constant definition in class.

- print_file_namesをリファクタリング
- formatted_detailメソッドをFileListerクラスからFileDetailクラスに移動
- 定数をクラス内に定義

sjabcdefin committed:
Update: Refactored calculation process in FileLister class.

FileListerクラスの計算処理をリファクタリング。

sjabcdefin committed:
Update: Modify class, method, variable name.

クラス名、メソッド名、変数名を変更。

yuuu commented on 07.ls_object/filedisplay.rb:
```diff
@@ -0,0 +1,84 @@
+require 'optparse'
+require_relative 'filedetail'
+
+class FileDisplay
```
`FileDisplay` というクラスの中には「列形式での出力」と「行形式での詳細出力」の処理が混在しており、それぞれの処理は関連がほとんどない状況です。これはオブジェクト指向の言葉で言うと「凝集度が低い」状態です。

クラス自体のボリュームも比較的大きいことを踏まえ、「列形式での出力」と「行形式での詳細出力」でクラスを分割してはどうでしょうか？

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/ruby-practices/pull/12/commits/accb4141358648ba4125d365544fc3aefb8484f7

```FileNameColumn```：列形式でのファイル名出力 と ```FileDetailRow```：行形式でのファイル詳細出力 にクラスを分割いたしました。
また、```FileDisplay```クラスの名前を```FileDisplayPreparation```に変更し、```FileDisplayPreparation```クラス内の```display_files```メソッド名を```display_files_based_on_format```に変更いたしました。

yuuu commented on 07.ls_object/filedisplay.rb:
```diff
@@ -0,0 +1,84 @@
+
+  def calculate_column_width
+    name_lengths = []
+    @file_details.each do |file_detail|
```
ここは `map` を使ったほうがシンプルに書けますね。

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/ruby-practices/pull/12/commits/15db0bc8742a1580012d5ded034b80d6254c3136

```map```に修正いたしました。

yuuu commented on 07.ls_object/filedisplay.rb:
```diff
@@ -0,0 +1,84 @@
+  def display_files
+    @file_details = @file_names.map { |file_name| FileDetail.new(file_name) }
+    if @options[:l_option]
+      max_size_length = calculate_max_size_length
```
一度変数に入れる意味がそこまでなさそうなので `formatted_detail` の引数にそのままメソッドを渡しても良さそうです 👍 
その場合、メソッドの名前を `max_link_length` としても良いですね。Rubyの場合、オブジェクト指向における `get` に相当する動詞はメソッド名から省略する文化があり、ここの `calculate` は `get` と同じニュアンスなので。

`max_link_length` についても同様です。

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/ruby-practices/pull/12/commits/e2128d76ab6e76f26170ad24a1c9a71fb38777a3

計算処理のメソッドについて、メソッド名から```calculate```を省略しました。
また、メソッドの戻り値を変数に設定していましたが、戻り値をそのまま使用するように修正いたしました。

yuuu commented on 07.ls_object/filedetail.rb:
```diff
@@ -0,0 +1,93 @@
+      group,
+      format("%#{max_size_length}d", size),
+      format('%<month>2d月 %<day>2d', update_time),
+      within_six_months? ? format('%<hour>02d:%<min>02d', update_time) : format('%<year>5d', update_time),
```
ここは [Time#strftime](https://docs.ruby-lang.org/ja/latest/method/Time/i/strftime.html) を使うと、 `update_time` のようなメソッドがいらなくなりコードもスッキリしそうですね 👍

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/ruby-practices/pull/12/commits/6e1a75d5538182c96901da8123f1e02b6b6c9ec0

最終更新日付のフォーマットに```strftime```を使用するように修正いたしました。
月と日が1桁の場合は空白埋めとするため、```%_m```, ```%e```を採用しています。
上記変更により、```update_time```メソッドが不要となったため、削除しました。

yuuu commented on 07.ls_object/filedisplay.rb:
```diff
@@ -0,0 +1,84 @@
+    interval = calculate_interval
+    file_names = @file_details.map(&:name)
+    file_names.each_slice(interval)
+              .to_a
```
後ろに続くメソッドが `map` であれば `Enumerable#map` が使えるので `.to_a` は省略ができますね。

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/ruby-practices/pull/12/commits/fd94d1f99e38c7b0e824156ccca4caa7995b338f

```.to_a```を削除（省略）いたしました。

sjabcdefin committed:
Refactor calculate_column_width method

@file_detailsの各要素からname_lengthを取得する処理について、以下を変更
- eachメソッドからmapメソッドに変更
- 冗長なブロックからシンプルなメソッドチェーンに変更

sjabcdefin committed:
Omit "calculate" from method name and Use the return value of the method as is

計算処理のメソッドについて、メソッド名から"calculate"を省略。
また、メソッドの戻り値を変数に設定していたが、そのまま使用するように修正。
- max_size_length
- max_link_length
- total_blocks
- interval
- column_width

sjabcdefin committed:
Refactor formatted_detail method for better date formatting and Delete update_time method

- 最終更新日付のフォーマットにstrftimeを使用するように変更
- 月と日が1桁の場合は空白埋めとするため、%_m, %eを採用
- 上記変更により、update_timeメソッドが不要となったため、削除

sjabcdefin committed:
Remove unnecessary to_a call in format_file_names_for_display method

each_sliceの結果は既にEnumeratorであり、Enumerable#mapを使用可能であったため、to_aを削除。

sjabcdefin committed:
Divide the class into "output in column format" and "detailed output in row format"

- 「行形式でのファイル詳細出力(FileDetailRow)」と「列形式でのファイル名出力(FileNameColumn)」でクラスを分割
- FileDisplayクラス名をFileDisplayPreparationクラスに変更
- FileDisplayPreparationクラスのdisplay_filesメソッド名をdisplay_files_based_on_formatに変更

yuuu commented on 07.ls_object/filedisplaypreparation.rb:
```diff
@@ -0,0 +1,37 @@
+require_relative 'filedetailrow'
+require_relative 'filenamecolumn'
+
+class FileDisplayPreparation
```
`FileDisplayPreparation` という命名ですが、このクラスで行っていることは「ファイルを出力する準備」でしょうか？

オプションを読み取って条件に合う形でファイルを見つけてきたり、出力形式を選択したり、これ自体がまさに `ls` のような気もしますがいかがでしょうか？

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/ruby-practices/pull/12/commits/e15f3d1fb16589ba4e5009b359d7f762b98d8fbf

ご認識の通り、「ファイルを出力する準備」のクラスとして、```FileDisplayPreparation```という命名にしました。
ご指摘いただいた通り、```ls```ですね！！！！😮
```FileDisplayPreparation```を```ls```に修正いたしました。
また、もともと、```ls```クラスがいましたので、そちらを```main```に修正いたしました。

yuuu commented on 07.ls_object/filedisplaypreparation.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  def display_files_based_on_format
+    file_display_handler = @options[:l_option] ? FileDetailRow.new(@file_details) : FileNameColumn.new(@file_details)
```
ここは `.new(@file_details)` も共通化ができそうですね。

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/ruby-practices/pull/12/commits/5f25da752f1c5300f77024abd3eb234b3f2974b4

```.new(@file_details)```も共通化いたしました。

yuuu commented on 07.ls_object/filedetailrow.rb:
```diff
@@ -0,0 +1,28 @@
+# frozen_string_literal: true
+
+class FileDetailRow
```
単語のチョイスとしては良い感じですが、クラス名はそのクラスが表現するモノの可算名詞単数形としたいです。このクラス名だと「1行」を1インスタンスが持つことになりますが、実際にはそうではないはずです。

`FileNameColumn` についても同様に見直しをお願いします 🙏

sjabcdefin replied:
```commit```：https://github.com/sjabcdefin/ruby-practices/pull/12/commits/86a15f657315ffc6be0664f920ac122409928aa1

```FileDetailRow```としていたクラスについては、「複数のファイル詳細情報を行形式に出力する」クラスであるため、```FileDetailDisplay```クラスに修正いたしました。
```FileNameColumn```としていたクラスについては、「複数のファイル名称を列形式に出力する」クラスであるため、```FileNameDisplay```クラスに修正いたしました。

「列形式」、「行形式」をどう表現するか悩みました。
ご指摘いただいた通り、```Row```、```Column```を使用すると1行、1列を指す印象を与えてしまう。。
表示形式である「列形式」、「行形式」を表す単語を省略し、「複数のファイル詳細情報を出力するクラス」、「複数のファイル名称を出力するクラス」としても、そのクラスが表現するモノを表現できているのではないかと考えました。
そのため、「列形式」、「行形式」を表す単語を省略する形で修正しております。🙇‍♀️

yuuu replied:
なるほど 👍 この命名も悪くはないと思うので、今回はこれで確定としましょう。
ちなみに、他の生徒で `Row` や `Column` といった単語を含めた名前を採用している方もいるはずなので、ぜひ他の生徒の提出物にも目を通して命名の引き出しを増やしてみてください 💪

sjabcdefin committed:
Change the class name and file name of filedisplaypreparation to ls and file name of ls to main

より適切なクラス名称とするために、filedisplaypreparation のクラス名とファイル名を ls に、ls のファイル名を main に変更

sjabcdefin committed:
Change the class name and file name of filedetailrow to filedetaildisplay and filenamecolumn to filenamedisplay

より適切なクラス名称とするために、filedetailrow のクラス名とファイル名を filedetaildisplay に、filenamecolumn を filenamedisplayに変更

sjabcdefin committed:
Refactor display_files_based_on_format method

よりシンプルなコードとするために、.new(@file_details) も共通化を実施
